text/metrics/0113-exemplars.md

```diff
@@ -0,0 +1,104 @@
+
+#### Last-Value
+
+Most (if not all) Last-Value aggregators operate asynchronously and do not ever interact with context. Since the value of a Last-Value is the last measurement (essentially the other parts of an exemplar), exemplars are not worth implementing for Last-Value.
+
+#### Exact
+
+The Exact aggregator will function by maintaining a list of `RawValue`s, which contain all of the information exemplars would carry. Therefore the Exact aggregator will not need to maintain any exemplars.
+
+#### Counter
+
+Exemplars give value to counter aggregations in two ways: One, by tying metric and trace data together, and two, by providing necessary information to re-create the input distribution. When enabled, the aggregator will retain a bounded list of exemplars at each checkpoint, sampled from across the distribution of the data. Exemplars should be sampled in a statistically significant way.
+
+#### MinMaxSumCount
+
+Similar to Counter, MinMaxSumCount should retain a bounded list of exemplars that were sampled from across the input distribution in a statistically significant way.
+
+#### Custom Aggregators
+
+Custom aggregators MAY support exemplars by maintaining a list of exemplars that can be retrieved by exporters. Custom aggregators should select exemplars based on their usage by the connected exporter (for example, exemplars recorded for Stackdriver should only be retained if they were recorded within a sampled trace).
+
+Exemplars will always be retrieved from aggregations (by the exporter) as a list of RawValue objects. They will be communicated via a
+
+```
+optional repeated RawValue exemplars = 6
+```
+
+attribute on the `Metric` object.
+
+## Trade-offs and mitigations
+
+Performance (in terms of memory usage and to some extent time complexity) is the main concern of implementing exemplars. However, by making recording exemplars optional, there should be minimal overhead when exemplars are not enabled.
+
+## Prior art and alternatives
+
+Exemplars are implemented in [OpenCensus](https://github.com/census-instrumentation/opencensus-specs/blob/master/stats/Exemplars.md#exemplars), but only for HistogramAggregator. This OTEP is largely a port from the OpenCensus definition of exemplars, but it also adds exemplar support to other aggregators.
+
+[Cloud monitoring API doc for exemplars](https://cloud.google.com/monitoring/api/ref_v3/rpc/google.api#google.api.Distribution.Exemplar)
+
+## Open questions
+
+- Exemplars usually refer to a span in a sampled trace. While using the collector to perform tail-sampling, the sampling decision may be deferred until after the metric would be exported. How do we create exemplars in this case?
+
+- We don’t have a strong grasp on how the sketch aggregator works in terms of implementation - so we don’t have enough information to design how exemplars should work properly.
```

Comment:
Apologies in advance if this is off-topic, but I'm trying to find a detailed spec on sketches in OT, and this is the closest I've found.

https://blog.acolyer.org/2019/09/06/ddsketch/ suggests at least one way that sketches can work, with code to implement found here: https://github.com/DataDog/datadog-agent/tree/master/pkg/quantile

The spec seems to be a bit vague about what the exemplar means in the case of sketches: is it the upper bound of the bucket?  the lower bound?   

Similarly, the content of the bucket, for histogram sketches, is ambiguous: are they cumulative (https://en.wikipedia.org/wiki/Histogram#Cumulative_histogram)?  For reference, Datadog histogram sketches are not cumulative, but Prometheus histogram sketches are cumulative (https://prometheus.io/docs/practices/histograms/#apdex-score).  It would be great if we could nail down these details in the spec, so that receivers and exporters can be implemented.

Reply:
Thank you @bsherrod. This is indeed not quite the topic of this OTEP, but it is very important and you are correct that there is something missing with regards to sketches. I am currently working on a few related matters. In the OTel-Go repository, there are several PRs associated with enabling exporters to choose whether they want cumulative or delta-oriented exports (the first attempt in a series here: https://github.com/open-telemetry/opentelemetry-go/pull/799). In the specification, I'm responsible for finishing the SDK specification--there's a WIP but it's out-of-date and doesn't answer the question about cumulative vs. delta (see the [WIP](https://github.com/open-telemetry/opentelemetry-specification/pull/347).

These matters ^^^ are mostly about specification, and I think we're pretty close to resolving them. There is something that we need to do in OTLP to support sending sketches, and then work to support sketches in the collector. I have done some research, but not enough research into good ways of encoding sketch data structures. In the OTel-Go repository, I've used the DDSketch implementation to implement approximate quantiles for Summary exports, but this isn't really how we want to export sketches.

There are bound to be many ways to encode sketches, and the protocol will want to address that. One mechanism that I favor is T-digest: the output of T-digest can be expressed as a set of weighted data points, which might be a useful representation--similar to a histogram but explicitly an approximate structure. A set of weighted data points is also similar in representation to a set of exemplars, so I guess your question is not really off topic after all. (I would propose that "set of exemplars" is a viable way to represent a sketch aggregation, but without an algorithm I can point to for computing this, I'll just leave this an open question.)

Reply:
I would invite you, if you're interested, to file an issue about this for OTLP in the https://github.com/open-telemetry/opentelemetry-proto repository.

Reply:
Thank you @jmacd !  I'll read the links you've added here and follow up.

Reply:
Re: Delta vs Cumulative aggregations, here's some progress: https://github.com/open-telemetry/opentelemetry-go/pull/840